src/test/groovy/com/streamr/client/SubscriptionSpec.groovy

```diff
@@ -0,0 +1,226 @@
+package com.streamr.client
+
+import com.streamr.client.options.ResendFromOption
+import com.streamr.client.options.ResendLastOption
+import com.streamr.client.protocol.control_layer.ResendRangeRequest
+import com.streamr.client.protocol.message_layer.StreamMessage
+import com.streamr.client.protocol.message_layer.StreamMessageV30
+import spock.lang.Specification
+
+class SubscriptionSpec extends Specification {
+
+    StreamMessageV30 msg = new StreamMessageV30("stream-id", 0, (new Date()).getTime(), 0, "publisherId", "msgChainId",
+            null, 0, StreamMessage.ContentType.CONTENT_TYPE_JSON, "{}", StreamMessage.SignatureType.SIGNATURE_TYPE_NONE, null)
+
+    boolean done = false
+    void setup() {
+        done = false
+    }
+
+    StreamMessage createMessage(long timestamp, long sequenceNumber, Long previousTimestamp, Long previousSequenceNumber) {
+        return createMessage(timestamp, sequenceNumber, previousTimestamp, previousSequenceNumber, "publisherId")
+    }
+
+    StreamMessage createMessage(long timestamp, long sequenceNumber, Long previousTimestamp, Long previousSequenceNumber, String publisherId) {
+        return new StreamMessageV30("stream-id", 0, timestamp, sequenceNumber, publisherId, "msgChainId",
+                previousTimestamp, previousSequenceNumber, StreamMessage.ContentType.CONTENT_TYPE_JSON, "{}", StreamMessage.SignatureType.SIGNATURE_TYPE_NONE, null)
+    }
+
+    MessageHandler empty = new MessageHandler() {
+        @Override
+        void onMessage(Subscription sub, StreamMessage message) {
+
+        }
+    }
+
+    void "calls the message handler"() {
+        when:
+        Subscription sub = new Subscription(msg.getStreamId(), msg.getStreamPartition(), new MessageHandler() {
+            @Override
+            void onMessage(Subscription sub, StreamMessage message) {
+                assert msg.toJson() == message.toJson()
+                done = true
+            }
+        })
+        sub.handleMessage(msg, 'session-token')
+        then:
+        done
+    }
+
+    void "calls the handler once for each message in order"() {
+        ArrayList<StreamMessage> msgs = new ArrayList<>()
+        for (int i=0;i<5;i++) {
+            msgs.add(createMessage((long)i, 0, null, 0))
+        }
+        ArrayList<StreamMessage> received = new ArrayList<>()
+        int counter = 0
+        when:
+        Subscription sub = new Subscription(msg.getStreamId(), msg.getStreamPartition(), new MessageHandler() {
+            @Override
+            void onMessage(Subscription sub, StreamMessage message) {
+                assert msgs.get(counter).toJson() == message.toJson()
+                received.add(message)
+                counter++
+                if (counter == 5) {
+                    done = true
+                }
+            }
+        })
+        for (int i=0;i<5;i++) {
+            sub.handleMessage(msgs.get(i), 'session-token')
+        }
+        then:
+        done
```

Comment:
Instead of the `done` stuff borrowed from JS world, we could assert here something like `received.equals(msgs)`